retention/cohort_analysis.py

```diff
@@ -222,9 +244,26 @@ def get_cohort_user_devices_bucket(users, start, stop):
     return res
 
 
-def get_user_agents(users, start):
+def get_user_agents(
+        users: Collection[str], start: int
+) -> Sequence[Tuple[str, str, Optional[str]]]:
+    """
+    Given a list of user ids, find the user agents of all the devices that they have
+    used to access the server since the given timestamp
+
+    Returns:
+        a list of (user_id, device_id, user_agent) triplets. Note that it may contain
+            duplicates as well as null user agents!
+    """
     if len(users) == 0:
         return []
+
+    # First we fetch a list of devices that the users have used since the given
+    # timestamp, and we then join that to both the `user_ips` and the `devices` tables
+    # to look for useragents that have come from that device.
+    #
+    # XXX: why do we check both tables? `devices` will only include one user-agent, but
```

Comment:
I'm really dredging my memory here, but I *think* the idea here is that `user_ips` only contains a more limited period (28 days?) worth of data than the `devices` table so by checking both we can provide more user-agents.